grpc.go

```diff
@@ -1000,29 +1001,31 @@ func isSignalHandled(pid int, signum syscall.Signal) bool {
 	return false
 }
 
-func (a *agentGRPC) WaitProcess(ctx context.Context, req *pb.WaitProcessRequest) (*pb.WaitProcessResponse, error) {
-	proc, ctr, err := a.sandbox.getProcess(req.ContainerId, req.ExecId)
-	if err != nil {
-		return &pb.WaitProcessResponse{}, err
-	}
-
-	defer proc.Do(func() {
-		proc.closePostExitFDs()
-		ctr.deleteProcess(proc.id)
-	})
-
-	// Using helper function wait() to deal with the subreaper.
+func (a *agentGRPC) waitProcess(proc *process, ctr *container) {
 	libContProcess := (*reaperLibcontainerProcess)(&(proc.process))
-	exitCode, err := a.sandbox.subreaper.wait(proc.exitCodeCh, libContProcess)
-	if err != nil {
-		return &pb.WaitProcessResponse{}, err
-	}
+	exitCode, _ := a.sandbox.subreaper.wait(proc.exitCodeCh, libContProcess)
+
 	//refill the exitCodeCh with the exitcode which can be read out
 	//by another WaitProcess(). Since this channel isn't be closed,
 	//here the refill will always success and it will be free by GC
 	//once the process exits.
 	proc.exitCodeCh <- exitCode
 
+	proc.closePostExitFDs()
+	ctr.deleteProcess(proc.id)
```

Comment:
You cannot delete the process here, since this fn would be called before the shim calling WaitProcess, and thus the shim wouldn't get the process's exit code.

Reply:
@lifupan thanks for you review. I think you are right ,but I  think shim can improve the call for WaitProcess, shim should not call WaitProcess after shim.proxyStdio, it's better to have them both in parallel(use go shim.proxyStdio https://github.com/kata-containers/shim/blob/master/main.go#L354, or other code in kata-containerd-shim-v2), this will make shim can get exitcode.  If  not parallel , Can we add some delay for closePostExitFDs?

Reply:
@keloyang 

Actually we shouldn't call the WaitProcess parallelly with shim.proxyStdio, since this would cause the lost of io data, what's why we called the WaitProcess after shim.proxyStdio done. 

For the issue of resource leak, I think it's not easy to avoid, since we depend on the shim to get the process's exit code, and if the shim didn't call the WaitProcess, then the essential resource should be kept for WaitProcess calling.

Reply:
@lifupan thanks for you review
>I think it's not easy to avoid

I think we can still do some thing to reduce the impact of the problem.  for example, when the process exit, kata-agent delay some time and then call deleteProcess,  so shim.Wait can get exitcode in the delay time. 

I have a usage scenario，if the exec process has no io data  in an hour, kata-proxy will close the connection from kata-shim.  this make shim.wait failed and lead kata-agent resource leak, at the end, kubectl can't exec to  the kata container.

I think the stability of the program should depend on itself as much as possible, normal operation should not cause program problems, just like our operating system, we constantly create a process and then kill it, this should not cause problems in our operating system.

Reply:
> @lifupan thanks for you review
> 
> > I think it's not easy to avoid
> 
> I think we can still do some thing to reduce the impact of the problem. for example, when the process exit, kata-agent delay some time and then call deleteProcess, so shim.Wait can get exitcode in the delay time.
> 
> I have a usage scenario，if the exec process has no io data in an hour, kata-proxy will close the connection from kata-shim. this make shim.wait failed and lead kata-agent resource leak, at the end, kubectl can't exec to the kata container.

Why the kata-proxy/kata-shim closed after no io data in an hour?  Did you killed it by your system? 

> 
> I think the stability of the program should depend on itself as much as possible, normal operation should not cause program problems, just like our operating system, we constantly create a process and then kill it, this should not cause problems in our operating system.

Actually, if you didn't reap/wait the child process, there would be zombie process left too.   It's the same with the agent issue. If you just create the process, and didn't do the reap, there's the case to left the resources in the system too.

Reply:
>Actually, if you didn't reap/wait the child process, there would be zombie process left too. It's the same with the agent issue. If you just create the process, and didn't do the reap, there's the case to left the resources in the system too.

@lifupan thanks， but I don't agree with you，create a process and then kill it , it is  a normal operation, even if don't reap it, the system init will reap the exited process. It is diffrent from kata-agent，if WaitProcess is not called，no other cat help kata-agent to do the closePostExitFDs. The leak resources  are not zombie  process, but fd.

I have add delay to call closePostExitFDs, PTAL, thanks.

Reply:
> > Actually, if you didn't reap/wait the child process, there would be zombie process left too. It's the same with the agent issue. If you just create the process, and didn't do the reap, there's the case to left the resources in the system too.
> 
> @lifupan thanks， but I don't agree with you，create a process and then kill it , it is a normal operation, even if don't reap it, the system init will reap the exited process. It is diffrent from kata-agent，if WaitProcess is not called，no other cat help kata-agent to do the closePostExitFDs. The leak resources are not zombie process, but fd.
> 

The init process only do the reap when the child process had no parent, if the parent process didn't exit and didn't do the wait/reap to the exited child process, then there would be zombie process left in the system too and the kernel/init wouldn't cleanup all of the process's resource for you. 

> I have added delay to call closePostExitFDs, PTAL, thanks.

The delay call cannot make sure the later wait call could get the process's exit code. I think you can add some resource cleanup here, but the some essential resources should be kept until the wait called to do the last resource clean.

Reply:
@lifupan thanks very much for your reply.

>The init process only do the reap when the child process had no parent

I understand the issue and when the exited process 's father become init, the init will reap the exited process, but I think it is not  the focus of our discussion. -_-
>add some resource cleanup here, but the some essential resources should be kept 

It is improvement， and I have changed th pr, please take a look.